content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2.md

```diff
@@ -0,0 +1,73 @@
+| File | Description |
+| -------- | -------- |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. **We map it to `requests.memory`.** |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. **We use a formula to calculate `memory.high` depending on `limits.memory/node allocatable memory` and a memory throttling factor.** |
+
+Both container and pod are considered. When cgroup v2 unified and MemoryQoS feature gate are both enabled, kubelet sets `requests.memory` to `memory.min` at container / pod level cgroup. And it also sets `memory.high` to throttle container memory overcommit allocation. 
+
+![](https://i.imgur.com/FpR0cUK.png)
+
+The formula is as follows:
+```
+// Container
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.min=pod.spec.containers[i].resources.requests[memory]
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.high=(pod.spec.containers[i].resources.limits[memory]/node allocatable memory)*memory throttling factor // Burstable
+// Pod
+/cgroup2/kubepods/pod<UID>/memory.min=sum(pod.spec.containers[i].resources.requests[memory])
+// QoS ancestor cgroup
+/cgroup2/kubepods/burstable/memory.min=sum(pod[i].spec.containers[j].resources.requests[memory])
+
+```
+
+Cgroup v2 memory controller would retain memory requests or throttle memory usages in the pod or container cgroup based on the set value. Because effective min boundary is limited by `memory.min` values of all ancestor cgroups, kubelet also sets qos or root cgroup if necessary.
+
+## How do I use it?
+All prerequisites must be met to enable the memory qos feature.
+
+1. Kubernetes since v1.22
+2. runc since v1.0.0-rc93, containerd since 1.14
+3. Linux kernel minimum version: 4.15, recommended version: 5.2+
+4. Kernel enables cgroups v2 unified hierarchy, add `systemd.unified_cgroup_hierarchy=1` to the kernel cmdline
+
+After that, user can enable memory qos in feature gate, add `--feature-gates="...,MemoryQoS=true"` to kubelet flag.
+
+## How can I learn more?
+
+You can find more details as follows:
+- KEPs
+    - Memory QoS https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos
+    - Cgroup v2 https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2
+- PRs
+    - Feature: add unified on CRI to support cgroup v2 https://github.com/kubernetes/kubernetes/pull/102578
+    - Feature: Support memory qos with cgroups v2 https://github.com/kubernetes/kubernetes/pull/102970
```

Comment:
```suggestion
     - [Support Memory QoS with cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos/#readme)
     - [cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2/#readme)
```
We usually _don't_ link directly from blogs about new features to PRs that are part of their implementation.